Enhanced relative layer visibility

CandyFace opened the pull request:
![reworked relativity](https://user-images.githubusercontent.com/1045397/53695627-37289d80-3dbe-11e9-9439-b82198127070.gif)

I have also implemented the ability to select the startup value for the global dot visibility mode.
![image](https://user-images.githubusercontent.com/1045397/53698752-479e3f80-3de1-11e9-8ff1-8189f6118d8c.png)
![image](https://user-images.githubusercontent.com/1045397/53698755-4cfb8a00-3de1-11e9-8d55-a777d80085f0.png)

fullfills and closes #1112

This PR adds:
+ Brings back relative layer visibility (see above)
  ~~- This brings back relative layer transparency with slight enhancements.
+ Left click to increase dot index, right click to decrease dot index
+ Ability to control relative visibility through preferences
+ Customizable layer visibility shortcuts
  - default shortcuts are alt+1, alt+2, alt+3

- [x]   Semi-Opaque (Transparent) Layers: Active layer is opaque. Passive Layers are transparent.
- [x]   Fully Opaque Layers: Active / Passive Layers are opaque by default
- [x]   Solo Layers: Active layer is opaque. Passive Layers are hidden from the user.
- [x]   Shortcut Key(s) Combination: Cycle Forwards / Cycle Backwards
- [x] Make fully opaque layer state default instead of semi-opaque / transparent state

~~Given the state of the timeline, I've left these out. I personally think that relative layer visibility should still be default, at least until it becomes apparent enough what the dots are used for (ie. to be changed via the timeline rewrite.)~~

- Move "Solo" State to Layer "visibility" manager

Not implemented exactly as requested:
- [x]   Widget Key Modifier: Inverse Cycling Order
You can cycle backwards using right-click now,

Jose-Moreno commented:
@CandyFace YES!!!!!!! Thank you so much! I know many people will love to have this back, myself included! :smile:

Jose-Moreno commented:
@CandyFace Alright I've read through all your notes. I just have one question. Will the visibility state be saved in a working pencil2d file? or will it be saved in the pencil2d preferences? 

I mean It's ok that transparent is default for a new user, but if the user presses alt +3 to make it fully opaque, will this change revert the next time they either open the program or open a new file?

CandyFace commented:
@Jose-Moreno The visibility state is not saved atm. so when you reopen a project it'll start as relative like it always has. I'll look into seeing if I can add that to the project file.

CandyFace commented:
@Jose-Moreno The visibility state is now being saved and loaded as requested.

CandyFace commented:
![reworked relativity](https://user-images.githubusercontent.com/1045397/53695627-37289d80-3dbe-11e9-9439-b82198127070.gif)

I've reworked the algorithm now so it goes both ways. So now it's "true" relative opacity based on the current layer in each direction as opposed to before where it only went downwards and the top layer would be the bottom of the stack and the only visible one.

Additionally I have also implemented the ability to select the startup value for the global dot visibility mode.
![image](https://user-images.githubusercontent.com/1045397/53698752-479e3f80-3de1-11e9-8ff1-8189f6118d8c.png)
![image](https://user-images.githubusercontent.com/1045397/53698755-4cfb8a00-3de1-11e9-8d55-a777d80085f0.png)


@scribblemaniac That's true I almost forgot about that.
Showing all layers fully opaque is now the default option.

CandyFace commented:
PR now up to take and everything still works.

CandyFace commented:
Finally got around updating the code after the other painting changes, can be reviewed now.

CandyFace commented on app/src/preferencesdialog.cpp:
```diff
@@ -314,6 +315,10 @@ TimelinePage::TimelinePage()
+    connect(ui->layerVisibilityComboBox, comboChanged, this, &TimelinePage::layerVisibilityChanged);
+    connect(ui->visibilitySlider, &QSlider::valueChanged, this, &TimelinePage::layerVisibilityThresholdChanged);
+    connect(ui->visibilitySpinbox, spinBoxValueChange, this, &TimelinePage::layerVisibilityThresholdChanged);
+    ui->visibilitySpinbox->setSuffix(" %");
```
Alright, will fix today.

CandyFace commented on app/src/preferencesdialog.cpp:
```diff
@@ -358,6 +363,19 @@ void TimelinePage::updateValues()
     ui->flipRollNumDrawingsSpinBox->setValue(mManager->getInt(SETTING::FLIP_ROLL_DRAWINGS));
     ui->flipInBtwnMsecSpinBox->setValue(mManager->getInt(SETTING::FLIP_INBETWEEN_MSEC));
+
+    qDebug() << mManager->getFloat(SETTING::LAYER_VISIBILITY_THRESHOLD);
```
No, I'll remove that

CandyFace commented on app/src/preferencesdialog.cpp:
```diff
@@ -380,6 +398,32 @@ void TimelinePage::playbackStateChanged(int value)
+        ui->visibilitySpinbox->setEnabled(true);
+    } else {
+        ui->visibilitySlider->setEnabled(false);
+        ui->visibilitySpinbox->setEnabled(false);
```
Hmm I always like that my UI is self explained, that is that if some logic is tied to some other element in the same view, then it should reflect that but maybe that's not so wise to do here. For consistency sake I'll remove it for now.

scribblemaniac replied:
A different way of looking at it is that the visibility percentage's relevance to the user is not determined by the default visibility mode, but rather the current visibility mode. Ex: the user may have hidden as the default but relative for the current project, and they would have to change their default to relative if they want to update the current visibility percentage.

CandyFace commented on app/ui/mainwindow2.ui:
```diff
@@ -225,6 +233,8 @@
     <addaction name="separator"/>
     <addaction name="menuChange_line_color"/>
+    <addaction name="separator"/>
+    <addaction name="menuOverall_Layer_Visibility"/>
```
Agreed, although a Light Table menu would probably be better in some future

CandyFace commented on app/ui/timelinepage.ui:
```diff
@@ -295,6 +295,72 @@
+             </widget>
+            </item>
+            <item>
+             <widget class="QSpinBox" name="visibilitySpinbox">
```
I've fixed it now, looks like some rounding error but I don't understand how that would stop the stepper...

Additionally that calculation was unnecessary since the value was already correct

CandyFace commented on core_lib/src/canvaspainter.cpp:
```diff
@@ -120,6 +120,9 @@ void CanvasPainter::initializePainter(QPainter& painter, QPixmap& pixmap)
     painter.setWorldTransform(mViewTransform);
+
+    QSettings settings(PENCIL2D, PENCIL2D);
+    mOpacityThreshold = settings.value(SETTING_LAYER_VISIBILITY_THRESHOLD).toReal();
```
To avoid more parameters, but i see now that, that was a silly decision.

CandyFace commented on core_lib/src/canvaspainter.cpp:
```diff
@@ -476,16 +479,25 @@ void CanvasPainter::paintTransformedSelection(QPainter& painter)
+    bool isCameraLayer = false;
+    if (mObject->getLayer(mCurrentLayerIndex)->type() == Layer::CAMERA) {
+        isCameraLayer = true;
+    }
```
Agreed

CandyFace commented on core_lib/src/canvaspainter.cpp:
```diff
@@ -131,7 +134,7 @@ void CanvasPainter::renderPreLayers(QPixmap *pixmap)
 void CanvasPainter::renderPreLayers(QPainter& painter)
 {
-    if (mOptions.nShowAllLayers > 0)
+    if (mOptions.elayerVisibility != CanvasPainterOptions::HIDDEN)
```
I personally think that goes against the UI expected behaviour, even if I agree that a solo camera layer is useless. I will make the change.

scribblemaniac replied:
It definitely goes against the expected behavior of the layer soloing, but I've always pushed the idea that when you're on the camera layer, you should see exactly what the camera sees and exactly what it will look like when exporting. So what we've got is a conflict between these two expected behaviors.

CandyFace commented on core_lib/src/canvaspainter.h:
```diff
@@ -31,6 +31,12 @@ class ViewManager;
+        HIDDEN = 0,
+        RELATIVE = 1,
+        FULL = 2
+    };
```
I will fix all cases except those that use signal/slot.. I don't see the reason to complicate signals, as long as the value is converted right after, that should be fine.

scribblemaniac replied:
> I don't see the reason to complicate signals, as long as the value is converted right after, that should be fine.

Agreed :+1:

CandyFace commented on core_lib/src/interface/scribblearea.cpp:
```diff
@@ -140,6 +149,9 @@ void ScribbleArea::settingUpdated(SETTING setting)
         mMultiLayerOnionSkin = mPrefs->isOn(SETTING::MULTILAYER_ONION);
         updateAllFrames();
         break;
+    case SETTING::LAYER_VISIBILITY_THRESHOLD:
```
Indeed I forgot it, why would I remove updateAllFrames? I still want the frame to update if the setting modified the frame.

scribblemaniac replied:
Look in setLayerVisibility :wink:

CandyFace replied:
Riiight... I guess it was that late, totally missed that XD

CandyFace commented on core_lib/src/interface/scribblearea.cpp:
```diff
@@ -1454,12 +1466,28 @@ void ScribbleArea::toggleOutlines()
+void ScribbleArea::increaseLayerVisibilityIndex()
+{
+    mLayerVisibility++;
+    if (mLayerVisibility == 3)
```
I figured out a decent way, simply convert from enum to enum and back to enum... but no it's not exactly pretty.

CandyFace commented on core_lib/src/managers/preferencemanager.cpp:
```diff
@@ -137,6 +139,12 @@ int PreferenceManager::getInt(SETTING option)
 
+float PreferenceManager::getFloat(SETTING option)
+{
+    float optionId = static_cast<int>(option);
```
whoops, that's a typo, will fix.

CandyFace commented on core_lib/src/structure/layer.cpp:
```diff
@@ -666,3 +666,24 @@ KeyFrame* Layer::getKeyFrameWhichCovers(int frameNumber) const
+        setId(id);
+    }
+    setName(elem.attribute("name"));
+    setVisible(elem.attribute("visibility").toInt());
```
Yes agreed

CandyFace commented on core_lib/src/structure/layercamera.cpp:
```diff
@@ -263,13 +263,9 @@ void LayerCamera::editProperties()
-    layerTag.setAttribute("visibility", visible());
-    layerTag.setAttribute("type", type());
-    layerTag.setAttribute("width", viewRect.width());
-    layerTag.setAttribute("height", viewRect.height());
```
Not really, I assume (hope) a typo

CandyFace commented on core_lib/src/structure/layer.cpp:
```diff
@@ -666,3 +666,24 @@ KeyFrame* Layer::getKeyFrameWhichCovers(int frameNumber) const
     return nullptr;
 }
+
+QDomElement Layer::createBaseDomElement(QDomDocument& doc)
```
👍

CandyFace commented on core_lib/src/canvaspainter.cpp:
```diff
@@ -495,6 +507,17 @@ void CanvasPainter::paintCurrentFrame(QPainter& painter, int startLayer, int end
+    int absoluteOffset = qAbs(layerOffset);
+    qreal newOpacity = 1.0;
+    if (absoluteOffset != 0) {
+        newOpacity = mOpacityThreshold/absoluteOffset;
```
Hmm you're right.. hah what an oversight...
Yes let's just go with pow for now, we can always introduce min and max sliders later.

scribblemaniac replied:
If we were Krita I bet there would probably be a curve editor for this setting :rofl:

CandyFace replied:
I wanted to bring back the layer transparency but in a way that resembled stacking paper on an actual light table light table. I know that Krita has a much more complex GUI where you can adjust the opacity for each layer but I really think that's overkill. A ranged slider could be made in the future for I think this suffice for now.

CandyFace commented on core_lib/src/canvaspainter.h:
```diff
@@ -108,6 +114,8 @@ class CanvasPainter : public QObject
     void paintAxis(QPainter& painter);
     void prescale(BitmapImage* bitmapImage);
 
+    /** @brief Calculate layer opacity based on current layer offset */
```
Noted, I've updated my config so that I don't need to remind myself of that :P

CandyFace replied:
I could, but why? There's no harm in it being there.. Do you want me to change it everywhere? I'll probably not do that.

scribblemaniac commented on app/ui/timelinepage.ui:
```diff
@@ -295,6 +295,72 @@
         </widget>
        </item>
+       <item>
+        <widget class="QGroupBox" name="LayerVisibilityBox">
```
Not resolved, LayerVisibilityBox and LayerVisibilityLayout remain.

CandyFace replied:
Weird.. LayerVisibilityLayout didn't show up in my Design editor... had to change it externally by modifying the xml directly.

scribblemaniac commented on core_lib/src/interface/scribblearea.cpp:
```diff
@@ -1454,12 +1462,28 @@ void ScribbleArea::toggleOutlines()
+void ScribbleArea::increaseLayerVisibilityIndex()
+{
+    mLayerVisibility = static_cast<LayerVisibility>(static_cast<int>(mLayerVisibility) + 1);
+    if (mLayerVisibility > LayerVisibility::FULL)
```
Well I would have converted to ints too, but this isn't quite the way I would go about doing this. In particular there are two issues I have with how this is currently implemented. First, it depends on LayerVisibility::Hidden being the first element and LayerVisibility::FULL being the last element, which is not clear elsewhere and is an odd requirement. Second, mLayerVisibility takes on a value outside of the defined values. To be honest I was surprised this worked at all, but apparently the specifications do allow for this. Still it makes me a little uneasy seeing this and imagining what were to happen if any other part of the program had to deal with these undefined values.

Here's what I would propose instead. Change the enum to this:
```c++
enum LayerVisibility {
    HIDDEN = 0,
    RELATIVE = 1,
    FULL = 2,
    COUNT = 3, // COUNT must always be last
};
```
Then the increaseLayerVisibilityIndex function changes to something like this:
```c++
mLayerVisibility = static_cast<LayerVisibility>((static_cast<int>(mLayerVisibility) + 1) % LayerVisibility::FULL);
updateAllFrames();
```
And decreaseLayerVisibilityIndex would be similar:
```c++
mLayerVisibility = static_cast<LayerVisibility>((static_cast<int>(mLayerVisibility) - 1 + LayerVisibility::FULL) % LayerVisibility::FULL);
updateAllFrames();
```
What do you think?

CandyFace replied:
I don't see what you achieve by doing this, if you think it's more safe, I disagree. The only difference here is you use count to iterate because you want to avoid an undefined variable that won't be used anyway and want to be sure the value you increment to matches one of the enums. I don't think is any better than incrementing to an undefined value which right after is reset so that the undefined behaviour will never be used it's untentional.

I've changed it though, adding two operators to increment/decrement within allowed range, no need for count still though.

scribblemaniac replied:
Well I disagree with your disagreeing :yum: . Based on my research ([mostly frome here](https://stackoverflow.com/questions/33812998/is-it-allowed-for-an-enum-to-have-an-unlisted-value#answer-33814842)), incrementing out of the enumeration range results in an unspecified value before C++17 and undefined behavior in C++17. That means the result could technically be any value so the assumption that `FULL+1 > FULL` is not good. When you switched to using a scoped enum this stopped being a problem because the underlying type is set to be int rather than whatever the compiler wants. The overrides are still a nice addition though :+1: 

I still think having COUNT would be useful to avoid accidentally forgetting to change the overrides when adding new enum value, but at least least the changes are close together now. I will add a comment to the code and leave it at that.

scribblemaniac commented on core_lib/src/structure/layer.cpp:
```diff
@@ -666,3 +666,24 @@ KeyFrame* Layer::getKeyFrameWhichCovers(int frameNumber) const
+        setId(id);
+    }
+    setName(elem.attribute("name", "untitled"));
+    setVisible(elem.attribute("visibility", "true").toInt());
```
The default value should be convertible to an int.

CandyFace replied:
Fixed. I expected a stringified bool to become an int but after testing I see that, that was a bad assumption.

scribblemaniac commented on core_lib/src/interface/scribblearea.cpp:
```diff
@@ -1119,18 +1126,19 @@ void ScribbleArea::prepCanvas(int frame, QRect rect)
+    o.bThinLines           = mPrefs->isOn(SETTING::INVISIBLE_LINES);
+    o.bOutlines            = mPrefs->isOn(SETTING::OUTLINES);
+    o.eLayerVisibility     = mLayerVisibility;
+    o.fLayerVisibilityThreshold = mPrefs->getFloat(SETTING::LAYER_VISIBILITY_THRESHOLD);
```
Well now you've got to realign everything with this :yum:

CandyFace replied:
It wasn't aligned before the pull request though, so I say let's leave it like that and do a proper refactoring later.

CandyFace commented on app/src/preferencesdialog.cpp:
```diff
@@ -358,6 +363,14 @@ void TimelinePage::updateValues()
     ui->flipRollNumDrawingsSpinBox->setValue(mManager->getInt(SETTING::FLIP_ROLL_DRAWINGS));
     ui->flipInBtwnMsecSpinBox->setValue(mManager->getInt(SETTING::FLIP_INBETWEEN_MSEC));
+
+    int convertedVisibilityThreshold = qFloor(static_cast<int>(mManager->getFloat(SETTING::LAYER_VISIBILITY_THRESHOLD)*100));
```
Technically you're right but that would leave an implicit conversion warning.

scribblemaniac replied:
Some time has past, perhaps you can look at this with fresh eyes now. You are taking a float, casting it to an int, implicitly converting it to a qreal (the type of the qFloor parameter), then you are flooring it back to an int.

In this context, flooring and truncating produce the exact same results. The only difference is when you are dealing with negative numbers. For the record: `qFloor(-0.5) == -1` but `static_cast<int>(-0.5) == 0`

CandyFace commented on core_lib/src/interface/timelinecells.cpp:
```diff
@@ -280,9 +280,9 @@ void TimeLineCells::drawContent()
-        if (mEditor->allLayers() == 2) { painter.setBrush(Qt::black); }
+        if (mEditor->layerVisibility() == LayerVisibility::HIDDEN) { painter.setBrush(Qt::NoBrush); }
+        if (mEditor->layerVisibility() == LayerVisibility::RELATIVE) { painter.setBrush(Qt::darkGray); }
+        if (mEditor->layerVisibility() == LayerVisibility::FULL) { painter.setBrush(Qt::black); }
```
Maybe but it's just going to be more lines, as long as it's only doing one thing, I'd prefer we leave it.
